### PR TITLE
Improve performance of ExtendedSchedulerPolicy

### PR DESCRIPTION
**Pull request by fviale**
- use an LRU cache for getStartAtValue (this prevents costly operations over the same job and task)
 - changed the key of delayedJobsWakeUpCache to be the startAt time instead of job id
 - added timer loggers to monitor performances when the logger is in debug

### CODE REVIEW

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java [hunk 1]
```diff
@@ -121,21 +138,28 @@ public LinkedList<EligibleTaskDescriptor> getOrderedTasks(List<JobDescriptor> jo
...
     }
 
-    private void scheduleWakeUp(final String taskOrJobId, Date now, Date startAtDate) {
-        if (!startAtCache.containsKey(taskOrJobId)) {
+    private void scheduleWakeUp(Date now, Date startAtDate) {
+        final long scheduledTime = startAtDate.getTime();
+        if (!delayedJobsWakeUpCache.containsKey(scheduledTime)) {
             executor.schedule(() -> {
-                startAtCache.remove(taskOrJobId);
+                delayedJobsWakeUpCache.remove(scheduledTime);
```
Comment (AlexFalessi) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java:153]: Not necessary as you are in `if (!delayedJobsWakeUpCache.containsKey(scheduledTime))`

**Review by AlexFalessi — APPROVED** [review]
Small comments

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java (hunk 1 above)
Comment (fviale) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java:153]: hehe, look at the code, it does not do it immediately, it schedules the operation.
When the operation is performed, the key must be removed

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSchedulingPolicy.java [hunk 2]
```diff
@@ -156,71 +164,81 @@ private void removeFinishedTasksAndReleaseTokens(String requiredLicense) {
...
                 return true;
+            } else { // == 0
+                logger.debug("Removing all terminated jobs/tasks and check again if there are remaining licenses");
+                removeFinishedJobsAndReleaseTokens(currentRequiredLicense);
+                removeFinishedTasksAndReleaseTokens(currentRequiredLicense);
+
+                if (licenseSynchronization.getRemainingTokens(currentRequiredLicense) > 0) {
+                    logger.debug("There are remaining licenses after releasing tokens, return true");
+                    return true;
+                }
```
[the remark this answers is not in the document]
Comment (fviale) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSchedulingPolicy.java:187]: I didn't touch this code, only wrapped it with timing loggers.

and yes I believe we can enter it if getRemainingTokens return 0

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java (hunk 1 above)
Comment (AlexFalessi) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/ExtendedSchedulerPolicy.java:153]: Oh yes sorry ...

**Pull request merged**